ir/src/depgraph.rs

```diff
@@ -0,0 +1,433 @@
+use filetime::FileTime;
+use serde::{Deserialize, Serialize};
+
+use std::{
+    collections::{HashMap, HashSet},
+    fs, io,
+    path::{Path, PathBuf},
+    vec,
+};
+
+/// Helps identify changes in a set of files.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(from = "DepGraphSerdeRepr", into = "DepGraphSerdeRepr")]
+pub struct DepGraph {
+    entries: HashMap<PathBuf, Option<DepGraphEntry>>,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct DepGraphEntry {
+    mtime: FileTime,
+    size: u64,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct DepGraphSerdeRepr {
+    entries: Vec<DepGraphEntrySerdeRepr>,
+}
+
+impl From<DepGraphSerdeRepr> for DepGraph {
+    fn from(from: DepGraphSerdeRepr) -> Self {
+        DepGraph {
+            entries: from
+                .entries
+                .iter()
+                .map(|d| {
+                    (
+                        PathBuf::from(&d.path),
+                        if d.has_graph_entry() {
+                            Some(DepGraphEntry {
+                                mtime: FileTime::from_unix_time(d.unix_seconds, d.nanos),
+                                size: d.size,
+                            })
+                        } else {
+                            None
+                        },
+                    )
+                })
+                .collect(),
+        }
+    }
+}
+
+impl From<DepGraph> for DepGraphSerdeRepr {
+    fn from(dg: DepGraph) -> Self {
+        DepGraphSerdeRepr {
+            entries: dg
+                .entries
+                .iter()
+                .map(|e| {
+                    let path = e.0.to_str().expect("Only UTF names please").to_string();
+                    if let Some(dep_entry) = e.1 {
+                        DepGraphEntrySerdeRepr {
+                            path,
+                            unix_seconds: dep_entry.mtime.unix_seconds(),
+                            nanos: dep_entry.mtime.nanoseconds(),
+                            size: dep_entry.size,
+                        }
+                    } else {
+                        DepGraphEntrySerdeRepr {
+                            path,
+                            unix_seconds: 0,
+                            nanos: 0,
+                            size: 0,
+                        }
+                    }
+                })
+                .collect(),
+        }
+    }
+}
+
+/// SystemTime lacks a platform independent representation we can
+/// depend on so use FileTime's unix_seconds,nanos.
+/// unix_seconds = nanos = size = 0 is used to represent lack of a DepGraphEntry.
```

Comment:
So it's worth keeping in mind how rustdoc renders doc comments. tl;dr, it's kind of like git commits: everything up until the first blank line is considered a 'summary', and is shown on index pages, where anything after the first line break is considered additional detail, and is only shown on the specific page for that item.

So basically: prefer a short summary sentence, and then further explanation below an empty (`/// `) line as needed.

More generally, though, this looks like it is just a comment and not really intended as public documentation?

Reply:
> just a comment and not really intended as public documentation?

Are doc comments of internals, aimed at maintainers, not a thing?

Reply:
Fixed the short/long pattern